### ISSUE TITLE
Soar V2: select and display WRF 2k map data automatically.

### ISSUE THREAD

**Issue by lhaeger** [soaringmeteo/soaringmeteo#164]
When switching between the three WRF 2k maps manually, the rest of the Alps does not display any forecast data on the map. Anyone living near the borders of the west/center/east maps for the WRF 2k forecasts will run into this, see [screen video ](https://youtu.be/JGXwXO3M2LM)for a demo.
To work around this I can open two browser tabs and display them side by side manually, but it would be great if Soar v2 could do that by itself and load & display all data matching the visible part of the map automatically.

In combination with #162 and #163 this would create a much smoother UI/UX.

**Comment by julienrf** [comment]
I understand your pain but I am not sure we could easily handle the areas where the zones overlap. Maybe there is an option to clip the projected image?
In the other hand, ultimately we would like to merge the zones together at the time we compute the WRF forwcast. As soon as we have enough computing power we will switch to that. When that will happen, that will solve your issue.

**Comment by lhaeger** [comment]
The front end basically downloads map tiles for coordinates, I assume. Why not merge the computed forecast results from the three independent runs, discarding the overlapping tiles from the center area? That should not require more computing resources but could be done as a final step before updating the web server. Maybe I just imaging it to be simple, though... ;-)

**Comment by lhaeger** [comment]
> we produce an image where every pixel represents a point of the grid (like [this one](https://github.com/user-attachments/assets/fabacf36-057b-4689-978f-97f076ee02de)), and we project the image on the map.

So you'd have to merge those three images into one, I assume.  Sounds doable, maybe even with a little [ImageMagick](https://usage.imagemagick.org/montage/). 
Would something similar be required for the wind direction and strength data?

**Comment by julienrf** [comment]
> So you'd have to merge those three images into one, I assume. Sounds doable

True, we might be able to merge the three images into a single one if the common grid points overlap perfectly. That would probably be simpler to do it on the server-side than on the client-side as I initially thought… Thanks for the suggestion!

> Would something similar be required for the wind direction and strength data?

Yes, and wind information is carried by vector tiles, so we would have to perform a similar merge.

**Comment by julienrf** [comment]
Would you be interested in helping with this task?

**Comment by lhaeger** [comment]
Sure, at least with the images I could give it a try if you can send me three example images that belong together (or attach them here). Look like the offsets/overlaps are:

between Central & Eastern Alps
- vertical offset (top border): 21px
- horizontal overlap: 12px

between Central & Southern Alps 
- vertical overlap: 18px
- horizontal offset (left border): 27px

I could try to merge them into a single image, with the parts not covered by the three inputs being transparent. Would that be what's needed?

What format do the vector tiles come in?

**Comment by julienrf** [comment]
Here are three images that should be mergeable together:

- [central](https://github.com/user-attachments/assets/e6ecd448-7463-489b-8911-bbc8b27c531b)
- [south](https://github.com/user-attachments/assets/59e5089c-2bdc-49b2-a375-e9a18cdd8fbb)
- [eastern](https://github.com/user-attachments/assets/601c5572-38d6-431c-a0eb-457e76607dec)

But please note that the overlapping pixels might not always have exactly the same color… This is because each zone gets its own forecast computation. So, you will probably have to look at the map to figure out which rows and columns overlap.

Currently, the backend generates one set of PNG files per WRF input file:

https://github.com/soaringmeteo/soaringmeteo/blob/7e81fdfbb7af279e307afae2213899faf41bc42b/backend/wrf/src/main/scala/org/soaringmeteo/wrf/DataPipeline.scala#L52

We would have to change that to group together some input files, and in the result we should return metadata with an extent that covers the three images.

The projection used to display the WRF result is defined here:

https://github.com/soaringmeteo/soaringmeteo/blob/7e81fdfbb7af279e307afae2213899faf41bc42b/frontend/src/map/Map.ts#L41

Regarding the vector tiles, the current implementation is here:

https://github.com/soaringmeteo/soaringmeteo/blob/main/backend/common/src/main/scala/org/soaringmeteo/out/VectorTiles.scala

In case you are familiar with raster tiles, the concept is similar in that we generate files corresponding to partitions of the space, but the difference is that those files contains vector information (in our case, points containing two properties: speed and direction).

Please let me know if you need any additional guidance!

**Comment by lhaeger** [comment]
Ok, finally found time to look into this. A single ImageMagick command can merge the three images into a single one covering the whole Alps:

`magick convert -size 456x252 xc:transparent -page +0+153 south.png -page +294+0 east.png -page +27+21 central.png -layers flatten alps.png`

![image](https://github.com/user-attachments/assets/ff115013-fb72-44bc-8648-154209c79f92)

I hope I got the image size and offsets correct, please double check:
 
[central.png](https://github.com/user-attachments/assets/e6ecd448-7463-489b-8911-bbc8b27c531b) : 279 × 150px, offset 27,21px
[south.png](https://github.com/user-attachments/assets/59e5089c-2bdc-49b2-a375-e9a18cdd8fbb) : size 129x99px, offset 0,153px
[eastern.png](https://github.com/user-attachments/assets/601c5572-38d6-431c-a0eb-457e76607dec) : size 162 × 144px, offset 294,0px

Resulting alps.png: size 456x252px

There are [quite a a few libraries available to call ImageMagick](https://imagemagick.org/script/develop.php) from code rather than command line, maybe that integrates better with your processes.

**Comment by lhaeger** [comment]
That approach sounds much better than trying to merge the current output, I completely agree. Let's hope you'll find the time to make it happen as I really do not think I can be of much help with it, unfortunately.

**Comment by julienrf** [comment]
Hi,
We are currently investigating whether we can mitigate this issue by increasing the overlap between the zones. Do you think the following zone would work better for pilots living around the border between the East zone and the Central zone?

![Image](https://github.com/user-attachments/assets/aa979d9c-d440-44ac-a122-ec023434d662)

For comparison, here is the current East zone:

![Image](https://github.com/user-attachments/assets/59b37877-b267-438a-a009-bed2cb126b0d)

You can also try it yourself here: https://soarwrf3.soaringmeteo.org/v2/?lat=47.348&lng=13.861&z=8.9&zone=eastern-alps

**Comment by lhaeger** [comment]
Hi Julien,

that overlap definitely makes sense, both for pilots from the Munich area (our main sites are now all covered by the eastern Alps zone) as well as people travelling to Bassano for a weekend, as it's now included in both the eastern and central zones.

But why this approach and not merge the data and then render it a single zone like you described in December?

Cheers, Lothar

**Comment by julienrf** [comment]
> But why this approach and not merge the data and then render it a single zone like you described in December?

Good question! Keeping the same zones but merging the data (option described in December) would probably provide a simpler user experience because users won’t have to think about the zones at all. However, the precision of the forecast is a bit worse at the zone border. Therefore, I think it is probably more interesting to make the zones bigger so that the areas of interest are less close to the zone border, and therefore the precision of the forecast is more reliable.

**Comment by lhaeger** [comment]
But that's not either-or, both could be combined, I think. You could discard the border tiles from both zones and still merge the remaining data into a single big picture. That improve ease of use without sacrificing accuracy. Double work, though...

**Comment by julienrf** [comment]
> Double work, though...

Yes :)
